Userland/Libraries/LibWeb/HTML/Navigable.cpp

```diff
@@ -379,8 +381,86 @@ Vector<JS::NonnullGCPtr<SessionHistoryEntry>>& Navigable::get_session_history_en
+
+        // 2. Return a clone of historyPolicyContainer.
+        return *history_policy_container;
+    }
+
+    // 2. If responseURL is about:srcdoc, then:
+    if (response_url == "about:srcdoc"sv) {
+        // 1. Assert: parentPolicyContainer is not null.
+        VERIFY(parent_policy_container.has_value());
+
+        // 2. Return a clone of parentPolicyContainer.
+        return *parent_policy_container;
+    }
+
+    // 3. If responseURL is local and initiatorPolicyContainer is not null, then return a clone of initiatorPolicyContainer.
+    if (Fetch::Infrastructure::is_local_url(response_url) && initiator_policy_container.has_value())
+        return *initiator_policy_container;
+
+    // 4. If responsePolicyContainer is not null, then return responsePolicyContainer.
+    // FIXME: File a spec issue to say "a clone of" here for consistency
+    if (response_policy_container.has_value())
+        return *response_policy_container;
+
+    // 5. Return a new policy container.
+    return {};
+}
+
+// https://html.spec.whatwg.org/multipage/browsers.html#obtain-coop
+static CrossOriginOpenerPolicy obtain_a_cross_origin_opener_policy(JS::NonnullGCPtr<Fetch::Infrastructure::Response>, Fetch::Infrastructure::Request::ReservedClientType const& reserved_client)
+{
+
+    // 1. Let policy be a new cross-origin opener policy.
+    CrossOriginOpenerPolicy policy = {};
+
+    // AD-HOC: We don't yet setup environments in all cases
+    if (reserved_client.has<Empty>())
+        return policy;
+
+    auto& reserved_environment = reserved_client.visit(
+        [](Empty const&) -> Environment& { VERIFY_NOT_REACHED(); },
+        [](Environment* env) -> Environment& { return *env; },
+        [](JS::GCPtr<EnvironmentSettingsObject> eso) -> Environment& { return *eso; });
+
+    // 2. If reservedEnvironment is a non-secure context, then return policy.
+    if (is_non_secure_context(reserved_environment))
+        return policy;
+
+    // FIXME: We don't yet have the technology to extract structured data from Fetch headers
+    // FIXME: 3. Let parsedItem be the result of getting a structured field value given `Cross-Origin-Opener-Policy` and "item" from response's header list.
+    // FIXME: 4. If parsedItem is not null, then:
+    //     FIXME: nested steps...
+    // FIXME: 5. Set parsedItem to the result of getting a structured field value given `Cross-Origin-Opener-Policy-Report-Only` and "item" from response's header list.
+    // FIXME: 6. If parsedItem is not null, then:
+    //     FIXME: nested steps...
+
+    // 7. Return policy.
+    return policy;
+}
+
+// https://html.spec.whatwg.org/multipage/browsing-the-web.html#attempt-to-create-a-non-fetch-scheme-document
+static JS::GCPtr<DOM::Document> attempt_to_create_a_non_fetch_scheme_document(NonFetchSchemeNavigationParams const&)
+{
+    // FIXME: Implement this algorithm to hand off to external software or display inline content
+    return nullptr;
```

Comment:
I would like to see FIXME message saying that we can't create document for the url if we reached this function.

Reply:
ah good point. now we print out that we don't know how to navigate to that URL to dbgln